brokers/artifacts/broker.go

```diff
@@ -58,122 +57,125 @@ func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) err
 	b.PubStarted()
 	b.PrintInfo("Starting plugin artifacts broker")
 
-	b.cleanupPluginsDirectory()
 	pluginMetas, err := utils.GetPluginMetas(pluginFQNs, defaultRegistry, b.ioUtils)
 	if err != nil {
 		return b.fail(fmt.Errorf("Failed to download plugin meta: %s", err))
 	}
-	b.PrintInfo("Downloading plugin extensions")
 
 	err = utils.ResolveRelativeExtensionPaths(pluginMetas, defaultRegistry)
 	if err != nil {
 		return b.fail(err)
 	}
 
-	for _, meta := range pluginMetas {
-		err = b.ProcessPlugin(meta)
+	requestedPlugins := convertMetasToPlugins(pluginMetas)
+
+	toInstall := b.SyncPlugins(requestedPlugins)
```

Comment:
It makes an impression that not every plugin that is requested should be installed.
But later I found 
```go
func (b *Broker) getPlan(requested, installed []model.CachedPlugin) (toInstall []model.CachedPlugin) {
	toInstall = requested
        ...
```
Maybe then `b.SyncPlugins(requestedPlugins)` should return nothing...
Or consumes Metas and returns `CachedPlugin` ?

Reply:
The idea behind `SyncPlugins` is that it syncs the list of requested plugins with what's on disk, i.e.
- Requested plugins should contain already-installed extensions
- Disk should not contain plugins that are not requested

I think the unclear thing is that the convention for `model.CachedPlugin` is that if a extension already has a path, it doesn't need to be downloaded. We set `toInstall = requested` since at the end of the day, all requested plugins must be installed, and then we partially fill `toInstall`'s extension urls with already downloaded paths.

If we make `b.SyncPlugins(requestedPlugins)` return nothing, then we basically move all the functionality in `broker.go` into `extensions.go`. Currently the division of labour is 
- extensions.go: keep up local disk cache of plugins
- broker.go: download extensions when necessary

As an example, suppose a workspace has `redhat/vscode-openshift-connector/latest` installed, and a Che update has caused `latest` to go from `0.0.21` to `0.1.0`. Running `SyncPlugins` will return
```go
model.CachedPlugin{
  ID: redhat/vscode-openshift-connector/latest
  IsRemote: true
  CachedExtensions: map[string]string {
    "https://github.com/[...]/vscode-kubernetes-tools-1.0.0.vsix": "<path to downloaded vsix>"
    "https://github.com/[...]/redhat.vscode-yaml-0.4.0.vsix": "<path to downlaoded vsix>"
    "https://github.com/[...]/openshift-connector-0.0.21-228.vsix": "" // empty string means should be downloaded
  }
}
```

Then `ProcessPlugin` will only download `https://github.com/[...]/openshift-connector-0.0.21-228.vsix` and update that entry in the map.

Reply:
Renamed method to `SyncWithPluginsDir`, hopefully that makes things a little clearer.

Reply:
> If we make b.SyncPlugins(requestedPlugins) return nothing, then we basically move all the functionality in broker.go into extensions.go. Currently the division of labour is
> 
> extensions.go: keep up local disk cache of plugins
> broker.go: download extensions when necessary

Just an idea to consider:

* broker.go: publish events, control downloading extensions and caching process by invoking the corresponding methods from files below
* extensions.go: download extensions when necessary
* cache.go: keep up local disk cache of plugins

Just minor comment what may be considered to be implemented in upcoming PRs.

Reply:
That makes a lot of sense, thanks.